core-backup/src/main/scala/io/aiven/guardian/kafka/backup/BackupClientInterface.scala

```diff
@@ -107,12 +149,101 @@ trait BackupClientInterface[T <: KafkaClientInterface] {
 
           Source.combine(
             Source.single((firstTimestamp, (firstReducedConsumerRecord, firstCursorContext))),
-            rest.map { case (reducedConsumerRecord, context) => (firstTimestamp, (reducedConsumerRecord, context)) }
+            rest.map { case (reducedConsumerRecord, context) =>
+              (firstTimestamp, (reducedConsumerRecord, context))
+            }
           )(Concat(_))
         case None => throw Errors.ExpectedStartOfSource
       }
     }
 
+  /** Transforms a sequence of [[RecordElement]]'s into a ByteString so that it can be persisted into the data storage
+    *
+    * @param sourceElements
+    *   A sequence of [[RecordElement]]'s as a result of `sliding(2)`
+    * @return
+    *   a [[ByteString]] ready to be persisted along with the original context form the [[RecordElement]]
+    */
+  private[backup] def transformReducedConsumerRecords(sourceElements: Seq[RecordElement]) = {
+    val stringWithContext = sourceElements match {
+      case Seq(Element(reducedConsumerRecord, context)) =>
+        // Happens in Sentinel case that is explicitly called at start of stream OR when a stream is interrupted by the user
+        // in which case stream needs to be terminated with `null]` in order to be valid
+        List((s"${reducedConsumerRecordAsString(reducedConsumerRecord)},", Some(context)))
+      case Seq(Element(firstReducedConsumerRecord, firstContext),
+               Element(secondReducedConsumerRecord, secondContext)
+          ) =>
+        List(
+          (s"${reducedConsumerRecordAsString(firstReducedConsumerRecord)},", Some(firstContext)),
+          (s"${reducedConsumerRecordAsString(secondReducedConsumerRecord)},", Some(secondContext))
+        )
+      case Seq(Element(reducedConsumerRecord, context), End) =>
+        List((s"${reducedConsumerRecordAsString(reducedConsumerRecord)}]", Some(context)))
+      case Seq(End) =>
+        List(("]", None))
+      case rest => throw Errors.UnhandledStreamCase(rest)
+    }
+    stringWithContext.map { case (string, context) => (ByteString(string), context) }
+  }
+
+  /** Applies the transformation to the first element of a Stream so that it starts of as a JSON array.
+    *
+    * @param element
+    *   Starting [[Element]]
+    * @param key
+    *   The current key being processed
+    * @param terminate
+    *   Whether to immediately terminate the JSON array for single element in Stream case
+    * @return
+    *   A [[List]] containing a single [[Start]] ready to be processed in the [[Sink]]
+    */
+  private[backup] def transformFirstElement(element: Element, key: String, terminate: Boolean) =
+    transformReducedConsumerRecords(List(element)).map {
+      case (byteString, Some(context)) =>
+        if (terminate)
+          Start(ByteString("[") ++ byteString.dropRight(1) ++ ByteString("]"), context, key)
```

Comment:
what's on that 1 byte that you drop?

Reply:
Its a `,` (comma), the `transformReducedConsumerRecords` adds a `comma` on the last character which is the standard case for a stream but this is a corner case where we immediately drop the `,` so that we have

```
...}]
```

instead of

```
..},]
```

(which is not valid JSON)

Reply:
either that or store this in a val with name `strippedTrailingComma`. That would be cool. I think this is one of the things that is quickly forgotten